app/scripts/modules/core/src/utils/RenderWhenVisible.tsx

```diff
@@ -0,0 +1,67 @@
+import * as React from 'react';
+
+export interface IRenderWhenVisibleProps {
+  // not valid without a container in px (IntersectionObserver doesn't do so good with null root but non-null rootMargin)
+  bufferHeight?: number;
+  // best guess to height of non-rendered content (in px)
+  placeholderHeight: number;
+  // set to false to call render eagerly
+  initiallyVisible?: boolean;
+  // standard render method; receives no props
+  render: () => JSX.Element;
+  // optional HTML element to use as the root in the IntersectionObserver
+  container?: Element;
+  // turns off de-rendering content
+  disableHide?: boolean;
+}
+
+export const RenderWhenVisible = ({
+  bufferHeight = 0,
+  initiallyVisible,
+  placeholderHeight,
+  render,
+  container,
+  disableHide,
+}: IRenderWhenVisibleProps) => {
+  const [isVisible, setIsVisible] = React.useState(!!initiallyVisible);
+  const [height, setHeight] = React.useState(placeholderHeight);
+  let observer: IntersectionObserver;
+
+  React.useEffect(() => {
+    return () => observer && observer.disconnect();
+  }, []);
+
+  const configureObserver = React.useCallback((node: HTMLDivElement) => {
+    if (!node) {
+      return;
+    }
+    let visible = isVisible;
+    observer = new IntersectionObserver(
+      entries => {
+        const inView = entries[0].isIntersecting;
+        if (inView && !visible) {
+          visible = true;
+          setIsVisible(true);
+        }
+        if (!inView && visible && !disableHide) {
+          visible = false;
+          setHeight(entries[0].boundingClientRect.height);
+          setIsVisible(false);
+        }
+      },
+      {
+        root: container,
+        threshold: [0],
+        rootMargin: bufferHeight && container ? `${bufferHeight}px 0px` : '0px',
+      },
+    );
+    observer.observe(node);
+  }, []);
+
+  return (
```

Comment:
Another way to do this is with refs and deps:

```
function Node() {
    const nodeRef = useRef();

    useEffect(() => {
      if (!nodeRef.current) {
        return;
      }

      console.log('Register observer')
      return () => console.log('Unregister observer')
    }, [nodeRef.current])
    return (
      <div ref={nodeRef}>
        Here's a node
      </div>
    );
}
```

Reply:
Updated to just `useEffect` with a `ref`. I was surprised in my testing to see I didn't need a null check on `nodeRef.current` - it's always there - and I didn't need to include `nodeRef.current` in my `deps`. The whole thing ended up a little cleaner.